doc/manual/src/design/store/objects.md

```diff
@@ -0,0 +1,73 @@
+# Store Objects
+
+Data in Nix is chiefly organized into *store objects*.
+A store object is the pair of
+
+  - A (root) file system object
+  - A set of references to store objects
+
+## File system objects
+
+The Nix store uses a simple filesystem model.
+
+    data FileSystemObject
+      = Regular Executable ByteString
+      | Directory (Map FileName FSO)
+      | SymLink ByteString
+
+    data Executable
+      = Executable
+      | NonExecutable
```

Comment:
```suggestion
    data FileSystemObject
      = File Executable Data
      | Directory (Map FileName FileSystemObject)
      | SymLink Path
```

I assume you've taken this from the thesis. Suggestion: Make more self-explanatory names, and skip the details for brevity.